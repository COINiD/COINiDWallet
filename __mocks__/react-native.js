import MockAsyncStorage from 'mock-async-storage';

import { WebSocket } from 'mock-socket';

const rn = require('react-native');

global.fetch = require('jest-fetch-mock');

global.fetch.mockResponse(JSON.stringify({ data: '12345' }));

global.WebSocket = WebSocket;

jest.useFakeTimers();

jest.mock('Alert', () => ({ alert: jest.fn() }));

jest.mock('AsyncStorage', () => new MockAsyncStorage());

jest.mock('NativeAnimatedHelper');

jest.mock('NativeModules', () => ({
  P2PTransferBLECentralModule: {
    addListener: jest.fn(),
  },
  P2PTransferBLEPeripheralModule: {},
  KeyboardObserver: {
    addListener: jest.fn(),
  },
  StatusBarManager: {
    HEIGHT: 42,
    setColor: jest.fn(),
    setStyle: jest.fn(),
    setHidden: jest.fn(),
    setNetworkActivityIndicatorVisible: jest.fn(),
    setBackgroundColor: jest.fn(),
    setTranslucent: jest.fn(),
  },
}));

module.exports = rn;
