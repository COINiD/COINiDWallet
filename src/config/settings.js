export default {
  coin : 'testnet',
  appReturnScheme : 'coinid-tbtc',
  coldWalletMode: true,
  language: 'system',
  availableLanguages : ['system','en','de','fr','es','pt','eo','sv'],
  currency: 'USD',
  availableCurrencies : ['USD','EUR','JPY','GBP','CAD','CHF','KRW','NZD','SEK','NOK','DKK'],
  ranges: ['day', 'week', 'month', 'year'],
  refreshRates: {
    exchangeRate: 600000, // 10 minutes
    account: 600000, // 10 minutes
  },
  coldTransportTypes: [
    {
      key: '',
      title: 'settings.coldtransport.showdialog.title',
      description: 'settings.coldtransport.showdialog.description',
    },
    {
      key: 'qr',
      title: 'settings.coldtransport.qrcodes.title',
      description: 'settings.coldtransport.qrcodes.description',
    },
    {
      key: 'ble',
      title: 'settings.coldtransport.ble.title',
      description: 'settings.coldtransport.ble.description',
    },
  ],
  lockDurations: [
    { milliseconds: 0, title: 'settings.lockduration.immediately' },
    { milliseconds: 60000, title: 'settings.lockduration.1minute' },
    { milliseconds: 300000, title: 'settings.lockduration.5minutes' },
    { milliseconds: 900000, title: 'settings.lockduration.15minutes' },
    { milliseconds: 1800000, title: 'settings.lockduration.30minutes' },
    { milliseconds: 3600000, title: 'settings.lockduration.1hour' },
    { milliseconds: 14400000, title: 'settings.lockduration.4hours' },
  ],
  aboutUrl: 'https://coinid.org',
  feedbackUrl: 'https://coinid.org/feedback',
  guidesUrl: 'https://coinid.org/guides',
  telegramUrl: 'https://t.me/joinchat/IARCoBAdhQOIEN_7u941Qg',
  twitterUrl: 'https://twitter.com/COINiDGroup',
  instagramUrl: 'https://www.instagram.com/coinidgroup',
  offlineGuideUrl: {
    ios: 'https://coinid.org/guides/preparing-ios-cold-wallet',
    android: 'https://coinid.org/guides/preparing-android-cold-wallet',
  },
  isTestnet : true,
};
