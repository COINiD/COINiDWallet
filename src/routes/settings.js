import { createStackNavigator } from 'react-navigation';
import SettingsRoute from '../components/SettingsRoute';

export const settingRoutes = {
  Home: {
    screen: SettingsRoute,
    swipeEnabled: false,
    title: 'Settings',
    isHome: true,
  },
  Passcode: {
    screen: SettingsRoute,
    title: 'Require unlocking',
  },
  OfflineTransport: {
    screen: SettingsRoute,
    title: 'Offline transport',
  },
  PreferredCurrency: {
    screen: SettingsRoute,
    title: 'Preferred currency',
  },
  SignMessage: {
    screen: SettingsRoute,
    title: 'Sign message',
  },
  About: {
    screen: SettingsRoute,
    title: 'About',
  },
  AccountList: {
    screen: SettingsRoute,
    title: 'Select an account',
  },
  AccountInformation: {
    screen: SettingsRoute,
    title: 'Account information',
  },
};

export const getSettingsNavigator = cardStyle => createStackNavigator(settingRoutes, {
  animationEnabled: true,
  headerMode: 'none',
  cardStyle,
});
