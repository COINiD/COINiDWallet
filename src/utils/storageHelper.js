/**
 * Helper for storage
 */

import { AsyncStorage } from 'react-native';

const getStorageItem = (namespace, key) => AsyncStorage.getItem(`@${namespace}:${key}`).then(JSON.parse);

const setStorageItem = (namespace, key, value) => {
  const val = JSON.stringify(value);
  return AsyncStorage.setItem(`@${namespace}:${key}`, val);
};

const removeStorageItem = (namespace, key) => AsyncStorage.removeItem(`@${namespace}:${key}`);

const reset = namespace => AsyncStorage.getAllKeys().then((keys) => {
  const filteredKeys = keys.filter(key => key.startsWith(`@${namespace}`));
  return AsyncStorage.multiRemove(filteredKeys);
});

export default namespace => ({
  get: key => getStorageItem(namespace, key),
  set: (key, value) => setStorageItem(namespace, key, value),
  remove: key => removeStorageItem(namespace, key),
  reset: () => reset(namespace),
});
