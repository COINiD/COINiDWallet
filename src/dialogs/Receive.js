import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet, View, TouchableOpacity, Platform,
} from 'react-native';
import Share from 'react-native-share';
import {
  DetailsModal, Text, AmountInput, ReceiveQRCode,
} from '../components';
import { colors, fontSize, fontWeight } from '../config/styling';
import ExchangeHelper from '../utils/exchangeHelper';
import ReceiveActionMenu from '../actionmenus/ReceiveActionMenu';

import WalletContext from '../contexts/WalletContext';

import styleMerge from '../utils/styleMerge';
import parentStyles from './styles/common';

const styles = styleMerge(
  parentStyles('light'),
  StyleSheet.create({
    smallText: {
      color: colors.getTheme('light').fadedText,
      margin: 0,
      width: '100%',
      ...fontWeight.medium,
    },
    amountForm: {
      borderBottomWidth: 1,
      borderBottomColor: colors.gray,
      flexDirection: 'row',
      paddingTop: 6,
    },
    amountInput: {
      color: colors.black,
      fontSize: fontSize.base,
      paddingBottom: 8,
      flex: 1,
      ...fontWeight.normal,
    },
    amountCurrency: {
      color: colors.getTheme('light').fadedText,
      paddingTop: Platform.OS === 'ios' ? 0 : 13,
      paddingLeft: 8,
    },
    currencyButton: {
      position: 'absolute',
      right: 0,
      bottom: 0,
      borderWidth: 1,
      borderColor: colors.purple,
      borderRadius: 8,
      paddingTop: 6,
      paddingBottom: 5,
      paddingHorizontal: 7,
      marginBottom: 8,
    },
    currencyButtonText: {
      color: '#617AF7',
      ...fontWeight.medium,
      letterSpacing: 0.1,
    },
  }),
);

class Receive extends PureComponent {
  static contextType = WalletContext;

  static propTypes = {
    address: PropTypes.string,
    theme: PropTypes.string,
  };

  static defaultProps = {
    address: '',
    theme: 'light',
  };

  constructor(props, context) {
    super(props);

    const { address } = props;
    const {
      coinid,
      globalContext: { settingHelper, showActionSheetWithOptions },
    } = context;
    const { ticker, coinTitle } = coinid;

    this.settingHelper = settingHelper;
    this.exchangeHelper = ExchangeHelper(ticker);

    this.state = {
      address,
      ticker,
      qrAddress: address,
      exchangeRate: 0,
      currency: '',
      amount: 0,
      coinTitle,
      showActionSheetWithOptions,
    };
  }

  getChildContext() {
    const { theme: propTheme } = this.props;
    const { theme: contextTheme } = this.context;

    return {
      theme: propTheme || contextTheme,
    };
  }

  componentDidMount() {
    const { address } = this.props;
    this._handleNewData({ address });

    this._onSettingsUpdated(this.settingHelper.getAll());
    this.settingHelper.on('updated', this._onSettingsUpdated);
  }

  componentWillReceiveProps(nextProps) {
    const { address: oldProps } = this.props;
    const { address } = nextProps;

    if (address !== oldProps) {
      this._handleNewData({ address });
    }
  }

  componentWillUnmount() {
    this.settingHelper.removeListener('updated', this._onSettingsUpdated);
  }

  _onSettingsUpdated = (settings) => {
    const { currency } = settings;
    this.setState({ currency });
    this._refreshExchangeRate(currency);
  };

  _refreshExchangeRate = (currency) => {
    this.exchangeHelper.convert(1, currency).then((exchangeRate) => {
      this.setState({ exchangeRate });
    });
  };

  _handleNewData = (newData) => {
    const { address: oldAddress, amount: oldAmount } = this.state;
    const { address = oldAddress, amount = oldAmount } = newData;

    const qrAddress = this._buildQrURI({ address, amount });
    this.setState({ qrAddress, address, amount });
  };

  _onChangeAmount = (amount) => {
    this._handleNewData({ amount });
  };

  _buildQrURI = ({
    address, amount, label, message,
  }) => {
    const {
      coinid: { qrScheme },
    } = this.context;
    let tmpl = [`${qrScheme}:`, address, '?']; // https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki

    if (amount) {
      tmpl = tmpl.concat(['amount=', encodeURIComponent(amount), '&']);
    }

    if (label) {
      tmpl = tmpl.concat(['label=', encodeURIComponent(label), '&']); // name of receiver
    }

    if (message) {
      tmpl = tmpl.concat(['message=', encodeURIComponent(message), '&']); // note
    }

    // Remove prefixing extra
    const lastc = tmpl[tmpl.length - 1];
    if (lastc === '&' || lastc === '?') {
      tmpl = tmpl.splice(0, tmpl.length - 1);
    }
    return tmpl.join('');
  };

  _open = () => {
    this.elModal._open();
  };

  _close = () => {
    this.elModal._close();
  };

  _validateAddress = () => {
    const { dialogNavigate } = this.context;
    const { address } = this.state;

    dialogNavigate(
      'ValidateAddress',
      {
        address,
      },
      this.context,
      false,
    );
  };

  _sweepPrivateKey = () => {
    const { dialogNavigate } = this.context;
    const { address } = this.state;

    dialogNavigate(
      'SweepPrivateKey',
      {
        address,
      },
      this.context,
    );
  };

  _share = async () => {
    const {
      address, amount, coinTitle, ticker,
    } = this.state;

    const getShareMessage = () => {
      let message = `My ${coinTitle} Address: ${address}`;

      if (amount) {
        message += `\nRequested Amount: ${amount} ${ticker}`;
      }

      return message;
    };

    const getQRUri = () => new Promise((resolve) => {
      this.viewShot.capture().then((uri) => {
        resolve(uri);
      });
    });

    const url = await getQRUri();

    const options = {
      title: 'Share via',
      message: getShareMessage(),
      url,
      type: 'image/png',
    };

    Share.open(options);
  };

  _showMoreOptions = () => {
    const { showActionSheetWithOptions } = this.state;

    const receiveActionMenu = new ReceiveActionMenu({
      showActionSheetWithOptions,
      onSweepPrivateKey: this._sweepPrivateKey,
      onValidateAddress: this._validateAddress,
      onShare: this._share,
    });

    receiveActionMenu.show();
  };

  _toggleInputFiat = () => {
    const { inputInFiat, exchangeRate } = this.state;

    if (!exchangeRate) {
      return false;
    }

    this.setState({ inputInFiat: !inputInFiat });
    return true;
  };

  render() {
    const {
      onOpened, onClosed, onOpen, onClose,
    } = this.props;

    const {
      ticker, qrAddress, address, exchangeRate, currency, amount,
    } = this.state;

    let { inputInFiat } = this.state;

    if (!exchangeRate) {
      inputInFiat = false;
    }

    return (
      <DetailsModal
        verticalPosition="flex-end"
        ref={(el) => {
          this.elModal = el;
        }}
        onOpened={onOpened}
        onClosed={onClosed}
        onOpen={onOpen}
        onClose={onClose}
        showMoreOptions
        onMoreOptions={this._showMoreOptions}
        avoidKeyboard
        avoidKeyboardOffset={40}
        title="Receive"
      >
        <View style={styles.container}>
          <View style={styles.modalContent}>
            <ReceiveQRCode
              getViewShot={(c) => {
                this.viewShot = c;
              }}
              address={address}
              qrAddress={qrAddress}
            />
          </View>

          <View
            style={styles.modalFooter}
            onLayout={(e) => {
              this.refContHeight = e.nativeEvent.layout.height;
            }}
          >
            <Text style={styles.smallText}>Request custom amount</Text>
            <View
              style={styles.amountForm}
              onFocus={(e) => {
                this.elModal._setKeyboardOffset(this.refAmountBottom - this.refContHeight + 8);
              }}
              onLayout={(e) => {
                this.refAmountBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height;
              }}
            >
              <AmountInput
                style={[styles.amountInput, { paddingRight: 60 }]}
                onChangeAmount={this._onChangeAmount}
                exchangeRate={exchangeRate}
                inputInFiat={inputInFiat}
                amount={amount}
                exchangeTo={currency}
                exchangeFrom={ticker}
              />
              <TouchableOpacity style={styles.currencyButton} onPress={this._toggleInputFiat}>
                <Text style={styles.currencyButtonText}>{inputInFiat ? currency : ticker}</Text>
              </TouchableOpacity>
            </View>
          </View>
        </View>
      </DetailsModal>
    );
  }
}

Receive.childContextTypes = {
  theme: PropTypes.string,
};

export default Receive;
