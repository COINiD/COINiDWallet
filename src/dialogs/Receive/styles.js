import { Platform, StyleSheet } from 'react-native';
import { colors, fontSize, fontWeight } from '../../config/styling';
import styleMerge from '../../utils/styleMerge';
import parentStyles from '../dialogStyles';

export default styleMerge(
  parentStyles('light'),
  StyleSheet.create({
    qrCode: {
      padding: 8,
      backgroundColor: colors.white,
    },
    qrCodeWrapper: {
      marginTop: 4,
      marginBottom: 16,
      height: 160,
      alignItems: 'center',
      flexDirection: 'row',
      justifyContent: 'center',
    },
    addressText: {
      fontSize: fontSize.small,
      textAlign: 'center',
    },
    smallText: {
      color: colors.getTheme('light').fadedText,
      margin: 0,
      width: '100%',
      ...fontWeight.medium,
    },
    amountForm: {
      borderBottomWidth: 1,
      borderBottomColor: colors.gray,
      flexDirection: 'row',
      paddingTop: 6,
    },
    amountInput: {
      color: colors.black,
      fontSize: fontSize.base,
      paddingBottom: 8,
      flex: 1,
      ...fontWeight.normal,
    },
    amountCurrency: {
      color: colors.getTheme('light').fadedText,
      paddingTop: Platform.OS === 'ios' ? 0 : 13,
      paddingLeft: 8,
    },
    currencyButton: {
      position: 'absolute',
      right: 0,
      bottom: 0,
      borderWidth: 1,
      borderColor: colors.purple,
      borderRadius: 8,
      paddingTop: 6,
      paddingBottom: 5,
      paddingHorizontal: 7,
      marginBottom: 8,
    },
    currencyButtonText: {
      color: '#617AF7',
      ...fontWeight.medium,
      letterSpacing: 0.1,
    },
  }),
);
