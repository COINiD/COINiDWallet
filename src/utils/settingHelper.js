'use strict';

/**
 * Helper for loading and saving settings.
 */

import { EventEmitter } from 'events';
import storageHelper from './storageHelper';
import Settings from '../config/settings';

class SettingHelper extends EventEmitter {
  constructor(coin) {
    super();

    this.setMaxListeners(50);

    this.storageNS = `${coin}:settings`;
    this.storage = storageHelper(this.storageNS);
    this.defaultSettings = {
      coldWalletMode: true,
      currencies: Settings.availableCurrencies,
      currency: Settings.currency,
      range: 0,
      usePasscode: true,
      lockAfterDuration: 60000,
      preferredColdTransport: '',
    };
    this.settings = { ...this.defaultSettings };
  }

  load = () => {
    const getDefaultIfEmpty = (key, data) =>
      data !== null ? data : this.defaultSettings[key];

    var p = Object.entries(this.defaultSettings).map(([key, value]) =>
      this.storage.get(key).then(data => getDefaultIfEmpty(key, data))
    );

    return Promise.all(p).then(
      ([
        coldWalletMode,
        currencies,
        currency,
        range,
        usePasscode,
        lockAfterDuration,
        preferredColdTransport,
      ]) => {
        this.settings = {
          coldWalletMode,
          currencies,
          currency,
          range,
          usePasscode,
          lockAfterDuration,
          preferredColdTransport,
        };

        this.emit('updated', this.settings);
      }
    );
  };

  update = (key, value) => {
    this.settings[key] = value;
    this.save();
    this.emit('updated', this.settings);
  };

  save = () => {
    var p = Object.entries(this.settings).map(([key, value]) => {
      return this.storage.set(key, value);
    });

    return Promise.all(p);
  };

  reset = () => {
    this.storage.reset();
  };

  getAll = () => {
    return this.settings;
  };
}

let settingHelpersCache = {}; // for local caching so we dont create several for same coin.

module.exports = function(coin) {
  if (settingHelpersCache[coin] === undefined) {
    settingHelpersCache[coin] = new SettingHelper(coin);
  }

  return settingHelpersCache[coin];
};
