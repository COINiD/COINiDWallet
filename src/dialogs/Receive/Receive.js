import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { View, Clipboard, TouchableOpacity } from 'react-native';
import QRCode from 'react-native-qrcode-svg';
import {
  DetailsModal, Text, FontScale, AmountInput,
} from '../../components';
import { MoreOptions, ValidateAddress } from '..';
import styles from './styles';
import { fontSize } from '../../config/styling';
import ExchangeHelper from '../../utils/exchangeHelper';

export default class Receive extends PureComponent {
  constructor(props, context) {
    super(props);

    const { address } = props;
    const { coinid, settingHelper } = context;
    const { ticker } = coinid;

    this.settingHelper = settingHelper;
    this.exchangeHelper = ExchangeHelper(ticker);

    this.state = {
      address,
      ticker,
      qrAddress: address,
      requestAmount: '',
      exchangeRate: 0,
      currency: '',
      amount: 0,
    };
  }

  getChildContext() {
    const { theme: propTheme } = this.props;
    const { theme: contextTheme } = this.context;

    return {
      theme: propTheme || contextTheme,
    };
  }

  componentDidMount() {
    const { address } = this.props;
    this._handleNewData({ address });

    this._onSettingsUpdated(this.settingHelper.getAll());
    this.settingHelper.on('updated', this._onSettingsUpdated);
  }

  componentWillReceiveProps(nextProps) {
    const { address: oldProps } = this.props;
    const { address } = nextProps;

    if (address !== oldProps) {
      this._handleNewData({ address });
    }
  }

  _onSettingsUpdated = (settings) => {
    const { currency } = settings;
    this.setState({ currency });
    this._refreshExchangeRate(currency);
  };

  _refreshExchangeRate = (currency) => {
    this.exchangeHelper
      .convert(1, currency)
      .then((exchangeRate) => {
        this.setState({ exchangeRate });
      });
  }

  _handleNewData = (newData) => {
    const { address: oldAddress, amount: oldAmount } = this.state;
    const { address = oldAddress, amount = oldAmount } = newData;

    const qrAddress = this._buildQrURI({ address, amount });
    this.setState({ qrAddress, address, amount });
  };

  _onChangeAmount = (amount) => {
    this._handleNewData({ amount });
  };

  _buildQrURI = ({
    address, amount, label, message,
  }) => {
    const { coinid: { qrScheme } } = this.context;
    let tmpl = [`${qrScheme}:`, address, '?']; // https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki

    if (amount) {
      tmpl = tmpl.concat(['amount=', encodeURIComponent(amount), '&']);
    }

    if (label) {
      tmpl = tmpl.concat(['label=', encodeURIComponent(label), '&']); // name of receiver
    }

    if (message) {
      tmpl = tmpl.concat(['message=', encodeURIComponent(message), '&']); // note
    }

    // Remove prefixing extra
    const lastc = tmpl[tmpl.length - 1];
    if (lastc === '&' || lastc === '?') {
      tmpl = tmpl.splice(0, tmpl.length - 1);
    }
    return tmpl.join('');
  }

  _open = () => {
    this.elModal._open();
  };

  _close = () => {
    this.elModal._close();
  };

  _copyAddress = () => {
    const { address } = this.state;
    Clipboard.setString(address);
    this.refMoreOptions._close();
  };

  _validateAddress = () => {
    const { address } = this.state;
    this.refMoreOptions._close();
    this.refValidateAddress._open(address);
  };

  _showMoreOptions = () => {
    this.refMoreOptions._open([
      [
        {
          title: 'Copy Address',
          onPress: this._copyAddress,
        },
        {
          title: 'Validate Address',
          onPress: this._validateAddress,
        },
      ],
      [
        {
          title: 'Cancel',
          onPress: this.refMoreOptions._close,
        },
      ],
    ]);
  };

  _toggleInputFiat = () => {
    const { inputInFiat, exchangeRate } = this.state;

    if (!exchangeRate) {
      return false;
    }

    this.setState({ inputInFiat: !inputInFiat });
  };

  render() {
    const {
      onOpened,
      onClosed,
      onOpen,
      onClose,
    } = this.props;

    const {
      ticker,
      qrAddress,
      address,
      exchangeRate,
      currency,
      amount,
    } = this.state;

    let { inputInFiat } = this.state;

    if (!exchangeRate) {
      inputInFiat = false;
    }

    return [
      <DetailsModal
        key="modal"
        verticalPosition="flex-end"
        ref={(el) => {
          this.elModal = el;
        }}
        onOpened={onOpened}
        onClosed={onClosed}
        onOpen={onOpen}
        onClose={onClose}
        showMoreOptions
        onMoreOptions={this._showMoreOptions}
        avoidKeyboard
        avoidKeyboardOffset={40}
        title="Receive"
      >
        <View
          style={styles.container}
        >
          <View style={styles.modalContent}>
            <View style={styles.qrCode}>
              <QRCode
                value={qrAddress}
                size={160}
                ecl="Q"
                logo={require('../../assets/images/qr_logo_full.png')}
                logoSize={80}
                logoBackgroundColor="transparent"
              />
            </View>
            <FontScale
              fontSizeMax={fontSize.small}
              fontSizeMin={8}
              text={address}
              widthScale={0.9}
            >
              {({ fontSize }) => (
                <Text style={[styles.addressText, { fontSize }]} selectable>
                  {address}
                </Text>
              )}
            </FontScale>
          </View>
          <View
            style={styles.modalFooter}
            onLayout={(e) => { this.refContHeight = e.nativeEvent.layout.height; }}
          >
            <Text style={styles.smallText}>Request custom amount</Text>
            <View
              style={styles.amountForm}
              onFocus={(e) => { this.elModal._setKeyboardOffset(this.refAmountBottom - this.refContHeight + 8); }}
              onLayout={(e) => { this.refAmountBottom = e.nativeEvent.layout.y + e.nativeEvent.layout.height; }}
            >
              <AmountInput
                style={[styles.amountInput, { paddingRight: 60 }]}
                onChangeAmount={this._onChangeAmount}
                exchangeRate={exchangeRate}
                inputInFiat={inputInFiat}
                amount={amount}
                exchangeTo={currency}
                exchangeFrom={ticker}
              />
              <TouchableOpacity
                style={styles.currencyButton}
                onPress={this._toggleInputFiat}
              >
                <Text style={styles.currencyButtonText}>
                  {inputInFiat ? currency : ticker}
                </Text>
              </TouchableOpacity>
            </View>
          </View>
        </View>
      </DetailsModal>,
      <MoreOptions key="more" ref={c => (this.refMoreOptions = c)} />,
      <ValidateAddress
        key="validateaddress"
        ref={c => (this.refValidateAddress = c)}
      />,
    ];
  }
}

Receive.contextTypes = {
  coinid: PropTypes.object,
  type: PropTypes.string,
  theme: PropTypes.string,
  settingHelper: PropTypes.object,
};

Receive.childContextTypes = {
  theme: PropTypes.string,
};

Receive.propTypes = {
  address: PropTypes.string,
  theme: PropTypes.string,
};

Receive.defaultProps = {
  address: undefined,
  theme: 'light',
};
