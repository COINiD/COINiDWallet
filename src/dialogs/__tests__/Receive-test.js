import 'react-native';
import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import PropTypes from 'prop-types';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
import Receive from '../Receive';

import COINiDPublic from '../../libs/coinid-public';
import settings from '../../config/settings';
import storageHelper from '../../utils/storageHelper';
import settingHelper from '../../utils/settingHelper';

const context = {
  coinid: COINiDPublic(
    settings.coin,
    storageHelper(`${settings.coin}-hot`),
    `${settings.coin}-hot`,
  ),
  settingHelper: settingHelper(settings.coin),
}

describe('Receive', () => {
  const defaultProps = {
    address: "msvEw9nCEtDtayUe9Mthh657yaDaqNN5A7",
  };

  const getWrapper = (props=defaultProps)Â => {
    const wrapper = shallow(
      <Receive {...props} />,
      { context },
    );

    const instance = wrapper.instance();
    instance.amountRef = { _updateAmount: jest.fn() };

    return { wrapper, instance };
  }

  it('gets matches snapshot', () => {
    const { wrapper, instance } = getWrapper();
    expect(wrapper.state()).toMatchSnapshot();
  });

});
