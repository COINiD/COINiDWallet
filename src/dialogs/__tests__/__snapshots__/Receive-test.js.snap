// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Receive gets matches snapshot 1`] = `
Object {
  "address": "msvEw9nCEtDtayUe9Mthh657yaDaqNN5A7",
  "amount": 0,
  "currency": "USD",
  "exchangeRate": 0,
  "qrAddress": "bitcoin:msvEw9nCEtDtayUe9Mthh657yaDaqNN5A7",
  "ticker": "tBTC",
}
`;
